crates/ra_editor/src/typing.rs

```diff
@@ -614,6 +637,31 @@ fn foo() {
         // ");
     }
 
+    #[test]
+    fn test_on_dot_typed() {
+        fn do_check(before: &str, after: &str) {
+            let (offset, before) = extract_offset(before);
+            let file = SourceFileNode::parse(&before);
+            let result = on_dot_typed(&file, offset).unwrap();
+            let actual = result.edit.apply(&before);
+            assert_eq_text!(after, &actual);
+        }
+        do_check(
+            r"
+    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
+        self.child_impl(db, name)
+        .<|>
+    }
+",
+            r"
+    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
+        self.child_impl(db, name)
+            .
+    }
+",
+        );
+    }
+
```

Comment:
A several more tests I think we should have is:

* don't reindernt if the dot is already indented. That is, edit should be None in this situation

```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .<|>
    }
```

* indent if the previous line is already indented

```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .first()
        .<|>
    }
```

->


```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .first()
            .<|>
    }
```

3) the combination of 1 & 2: no indent if dot is already indented and there's an indent on the previous line.